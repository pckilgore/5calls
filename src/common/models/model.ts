import { Issue, Category } from './issue';

export interface Outcome {
  label: string;
  status: string;
}

export interface VoterContact {
  id: string;
  name: string;
  location: string;
  phone: string;
}

export class CategoryMap {
  category: Category;
  issues: Issue[];
}

export type Party = 'democrat' | 'republican' | 'independent' | '';

export interface FieldOffice {
  city: string;
  phone: string;
}

export interface UserStat {
  all: string[];
  contactedCount: number;
  voiceMailCount: number;
  unavailableCount: number;
}

/**
 * Represents the place used to get location.
 * It may be one of three options:
 * 1. CACHED_ADDRESS - address stored in local storage
 * 2. BROWSER_GEOLOCATION - address obtained from the browser geolocation API
 * 3. IP_INFO - address obtained from ipinfo.io API
 */
export enum LocationFetchType {
  CACHED_ADDRESS = 'CACHED_ADDRESS', // 'address' in Choo version
  BROWSER_GEOLOCATION = 'BROWSER_GEOLOCATION', // 'browserGeolocation' in Choo version
  IP_INFO = 'IP_INFO' // 'ipAddress' in Choo version
}

/**
 * Encapsulates a location using latitude
 * and longitude. Undefined for either longitude
 * or latitude indicates that the geolocation has
 * has not been set.
 */
export interface GeolocationPosition {
  longitude: number | undefined;
  latitude: number | undefined;
}

/* 5 Calls API data */
export interface IssueData {
  issues: Issue[];
}

export interface GroupIssues {
  splitDistrict: boolean;
  invalidAddress: boolean;
  normalizedLocation: string | undefined;
  divisions: string[];
  issues: Issue[];
}

export interface CountData {
  count: number; // total call count
}

/* Data from iponfo.io API */
export interface IpInfoData {
  ip: string;
  hostname: string;
  city: string;
  region: string; // state
  country: string;
  loc: string; // long, lat - used in issue lookup
  org: string; // internet service provider
  postal: string; // zip code
}

export interface DonationGoal {
  goal: Donations; //
}

export interface Donations {
  count: number; // number of donors
  amount: number; // total collected (api===amount)
  total: number; // goal (api===total)
  kind: string; // denomincation (dollars)
}

export interface OutcomeButton {
  title: string;
  emoji: string;
  key: string;
}
