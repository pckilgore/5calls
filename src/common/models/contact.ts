import { Party, FieldOffice } from './model';

export class Contact {
  id: string;
  name: string;
  phone: string;
  photoURL?: string;
  party: Party;
  state: string;
  reason: string;
  area?: string;
  // tslint:disable-next-line:variable-name
  field_offices?: FieldOffice[];

  public contactDisplay(): string {
    if (this.partyStateAbbr() !== '') {
      return this.name + ' (' + this.partyStateAbbr() + ')';
    }

    return this.name;
  }

  public partyStateAbbr(): string {
    if (this.party && this.state !== '') {
      const partyLetter = this.party.substring(0, 1);
      return partyLetter.toUpperCase() + '-' + this.state.toUpperCase();
    }

    return '';
  }
}

export const mockContact = Object.assign(new Contact(), {
  id: 123,
  name: 'Mock Contact III',
  phone: '4155551212',
  photoURL: undefined,
  party: 'Democrat',
  state: 'CA',
  reason: 'This is your mock representative',
  area: undefined,
  field_offices: []
});
