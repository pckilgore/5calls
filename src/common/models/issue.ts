import { Outcome, Contact, ContactList } from './';

export class Issue {
  public id: number;
  public name: string;
  public contactType: string;
  public contactAreas: string[];
  public reason: string;
  public script: string;
  public categories: Category[];
  public active: boolean;
  public createdAt: string;
  public slug: string;
  public outcomeModels: Outcome[];
  public link: string;
  public linkTitle: string;
  public stats: IssueStats;

  public slugOrID(): string {
    if (this.slug !== '') {
      return this.slug;
    }

    return this.id.toString();
  }

  public numberOfContacts(contactList: ContactList): number {
    return this.filteredContacts(contactList).length;
  }

  public currentContact(
    contactList: ContactList,
    contactIndex: number
  ): Contact | undefined {
    const contacts = this.filteredContacts(contactList);

    if (contactIndex <= contacts.length) {
      return contacts[contactIndex];
    }

    return undefined;
  }

  // why do we need filtered contacts if we already have an array of contacts?
  // a) ensuring consistent ordering
  // b) sometimes we get back multiple house reps in split areas
  public filteredContacts(contactList: ContactList): Contact[] {
    const contacts: Contact[] = [];

    for (const contactArea of this.contactAreas) {
      if (contactArea === 'US Senate') {
        contacts.push(...contactList.senateReps());
      } else if (contactArea === 'US House') {
        contacts.push(...contactList.houseRep());
      } else if (contactArea === 'Governor') {
        contacts.push(...contactList.governor());
      } else if (contactArea === 'StateUpper') {
        contacts.push(...contactList.stateUpper());
      } else if (contactArea === 'StateLower') {
        contacts.push(...contactList.stateLower());
      } else if (contactArea === 'SecretaryOfState') {
        contacts.push(...contactList.secState());
      } else if (contactArea === 'AttorneyGeneral') {
        contacts.push(...contactList.attyGeneral());
      }
    }

    return contacts;
  }

  public listItemLabel(): string {
    switch (this.contactType) {
      case 'REPS':
        return 'Call reps' + this.countLabel();
      case 'ACTION':
        return 'Take action' + this.countLabel();
      default:
        return '???';
    }
  }

  public countLabel(): string {
    if (this.stats.calls < 10) {
      return '';
    }

    switch (this.contactType) {
      case 'REPS':
        return ' ⋆ ' + this.stats.calls + ' calls made';
      case 'ACTION':
        return ' ⋆ ' + this.stats.calls + ' actions taken';
      default:
        return '???';
    }
  }
}

export interface IssueStats {
  completion: number;
  calls: number;
}

export interface Category {
  name: string;
}

export const mockIssue = Object.assign(new Issue(), {
  id: 123,
  name: 'Issue Name',
  contactType: 'REPS',
  contactAreas: [],
  reason: 'reason',
  script: 'issue *mock script*',
  active: true,
  createdAt: '',
  slug: 'issue-name',
  outcomeModels: [],
  link: '',
  linkTitle: '',
  stats: { completion: 0, calls: 10 }
});
