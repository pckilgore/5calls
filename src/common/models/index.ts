import { Issue, Category } from './issue';
import { Contact } from './contact';
import { ContactList } from './contactList';
import {
  Outcome,
  OutcomeButton,
  LocationFetchType,
  FieldOffice,
  CategoryMap,
  GeolocationPosition,
  CountData
} from './model';

export {
  Issue,
  Category,
  Contact,
  ContactList,
  Outcome,
  OutcomeButton,
  LocationFetchType,
  FieldOffice,
  CategoryMap,
  GeolocationPosition,
  CountData
};
