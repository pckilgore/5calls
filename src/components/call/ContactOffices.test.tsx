import * as React from 'react';
import { shallow } from 'enzyme';
import i18n from '../../services/i18n';
import { ContactOffices } from './index';
import { mockContact } from '../../common/models/contact';
import { mockIssue } from '../../common/models/issue';

// these were just to match the snapshots previously, which is shit for testing these things
test('Contact offices should display if available', () => {
  const component = shallow(
    <ContactOffices currentIssue={mockIssue} currentContact={mockContact} />
  );
  expect(component).toMatchSnapshot();
});

test('Contact offices should not display if unavailable', () => {
  const component = shallow(
    <ContactOffices currentIssue={mockIssue} currentContact={mockContact} />
  );
  expect(component).toMatchSnapshot();
});
