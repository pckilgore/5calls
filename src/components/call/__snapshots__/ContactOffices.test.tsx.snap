// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact offices should display if available 1`] = `<span />`;

exports[`Contact offices should not display if unavailable 1`] = `<span />`;
